### PR TITLE
Add registry cache garbage collection for serverless

### PR DESCRIPTION
**Pull request by moelsayed**
**Description**

Changes proposed in this pull request:

- Add registry cache garbage collection
- ...
- ...

**Related issue(s)**

[16208](https://github.com/kyma-project/kyma/issues/16208)

### CODE REVIEW

**Comment by anoipm** [comment]
It's ok for me. I'm waiting for `pull-function-controller-unit-test` fix.

**Review by anoipm — DISMISSED** [review]

File: components/function-controller/design/internal_registry_garbage_collection.md [hunk 1]
```diff
@@ -5,7 +5,9 @@
 
 The current implementation for the Internal Registry deployed with Kyma Functions doesn't include any garbage collection logic. In dynamic environments with several Functions with multiple versions, if the Internal Registry is used with a cluster PVC storage backend, the volume disk space can fill up very quickly, resulting in build failures for new Function versions.
 
-Alternatively, if cloud storage like S3 is used as a storage backend, the old image blobs will increase S3 cost overtime.
+In addition the the space consumed by the Function images, the Function build process pushes the build cache layers to registry for later faster builds.
```
Comment (grego952) [inline components/function-controller/design/internal_registry_garbage_collection.md:8]:
```suggestion
In addition to the space consumed by the Function images, the Function build process pushes the build cache layers to the registry for later, faster builds.
```

File: components/function-controller/design/internal_registry_garbage_collection.md [hunk 2]
```diff
@@ -25,22 +27,32 @@ To utilize this, we need to implement custom logic that will periodically run an
 - Identify current images used by those Functions' runtimes.
 - List all tags for the used images.
 - Identify unused tags and use the Registry API to delete them.
+- List all cache layers on the registry
```
Comment (grego952) [inline components/function-controller/design/internal_registry_garbage_collection.md:30]:
```suggestion
- List all cache layers on the registry.
```

File: components/function-controller/design/internal_registry_garbage_collection.md [hunk 3]
```diff
@@ -25,22 +27,32 @@ To utilize this, we need to implement custom logic that will periodically run an
...
 
 - Identify existing Functions running on the cluster.
 - Identify current images used by those Functions' runtimes.
 - List all tags for the used images.
-- Identify unused tags and use the Registry API to delete them
+- Identify unused tags and use the Registry API to delete them.
+
+#### Cache image garbage collection
+- List all non-cache layers on the Registry. This is done after applying the Function image garbage collection to make sure only referenced images are listed.
+- List all the cache layers. Each layer is mapped to it's referencing tag.
```
Comment (grego952) [inline components/function-controller/design/internal_registry_garbage_collection.md:53]:
```suggestion
- List all the cache layers. Each layer is mapped to its referencing tag.
```

**Review by grego952 — APPROVED** [review]

**Review by anoipm — APPROVED** [review]

**Pull request merged**